pkg/apk/repo_test.go

```diff
@@ -142,66 +166,91 @@ func TestGetRepositoryIndexes(t *testing.T) {
-		require.NoError(t, err, "unable to read previous index file")
 		require.Equal(t, index1, index2, "index files do not match")
 	})
 	t.Run("cache hit etag miss", func(t *testing.T) {
-		// it should get the new one for a cache hit but etag mismatch
+		// it should succeed for a cache hit
 		tmpDir := t.TempDir()
-		// fill the cache
-		repoDir := filepath.Join(tmpDir, url.QueryEscape(testAlpineRepos), testArch)
-		err := os.MkdirAll(repoDir, 0o755)
-		require.NoError(t, err, "unable to mkdir cache")
-		index, err := os.ReadFile(filepath.Join(testPrimaryPkgDir, indexFilename))
-		require.NoError(t, err, "unable to read index")
-		cacheIndexFile := filepath.Join(repoDir, indexFilename)
-		err = os.WriteFile(cacheIndexFile, index, 0o644) //nolint:gosec // we're writing a test file
-		require.NoError(t, err, "unable to write index")
 		testEtag := "test-etag"
-		err = os.WriteFile(filepath.Join(repoDir, indexFilename+".etag"), []byte(testEtag), 0o644) //nolint:gosec // we're writing a test file
-		require.NoError(t, err, "unable to write etag")
+
 		// get our APK struct
 		a := prepLayout(t, tmpDir)
+
 		a.SetClient(&http.Client{
-			// since the etag is unchanged, our final file should be the same as the original
-			// the one in testdata/alpine-317 is different
-			Transport: &testLocalTransport{root: testAlternatePkgDir, basenameOnly: true, headers: map[string][]string{http.CanonicalHeaderKey("etag"): {testEtag + "abcdefg"}}},
+			Transport: &testLocalTransport{root: testPrimaryPkgDir, basenameOnly: true, headers: map[string][]string{http.CanonicalHeaderKey("etag"): {testEtag}}},
 		})
+		// Use the client to fill the cache.
 		indexes, err := a.getRepositoryIndexes(context.TODO(), false)
 		require.NoErrorf(t, err, "unable to get indexes")
 		require.Greater(t, len(indexes), 0, "no indexes found")
+		// Capture the initial index.
+		index1 := indexes[0]
+
+		// Update the transport to serve a different etag and different content,
+		// to verify that when the etag changes we use the data from the
+		// response.
+		a.SetClient(&http.Client{
+			Transport: &testLocalTransport{root: testAlternatePkgDir, basenameOnly: true, headers: map[string][]string{http.CanonicalHeaderKey("etag"): {testEtag + "change"}}},
+		})
+		indexes, err = a.getRepositoryIndexes(context.TODO(), false)
+		require.NoErrorf(t, err, "unable to get indexes")
+		require.Greater(t, len(indexes), 0, "no indexes found")
+		// Capture the resulting index.
+		index2 := indexes[0]
+
 		// check that the contents are the same
-		index1, err := os.ReadFile(cacheIndexFile)
-		require.NoError(t, err, "unable to read cache index file")
-		index2, err := os.ReadFile(filepath.Join(testAlternatePkgDir, indexFilename))
-		require.NoError(t, err, "unable to read previous index file")
-		require.Equal(t, index1, index2, "index files do not match")
+		require.NotEqual(t, index1, index2, "index files do not match")
+	})
+	t.Run("test cache concurrency", func(t *testing.T) {
```

Comment:
This test fails immediately (over and over, as we have been seeing) at HEAD.

With this change it passes.